plutus-common/src/Language/Plutus/Common.hs

```diff
@@ -0,0 +1,73 @@
+pir compiling, plc executing, "offline" runtime plutus code ---, so
+as to document and easily identify these plutus errors.
+
+We drew inspiration from `rustc` compiler error-codes:
+<https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/index.html>
+
+An errorcode is a positive number (`Word`) assigned to every possible data-constructor
+that represents an exceptional case. This includes both pure error-values raised
+by e.g. `ExceptT` but also "impure" ghc-builtin Control.Exception instances.
+
+For that we created a class `ErrorCode` with one method `errorCode`,
+left to be implemented for each error by the Plutus developers.
+It is the responsibility of the  Plutus developer to make sure that
+
+1) the assigned errorcode (Word) is unique among the whole Plutus-project codebase,
+2) the `errorCode` method is total
+3) no "wrapper-constructors" are tagged. e.g.in:
+
+```data PirError =
+    WrapperTC PIR.TCError
+    | WrapperParse  PIR.ParseError
+    | PirCompile String
+```
+
+we do not uniquely tag the wrapper-constructors WrapperTC,WrapperParse,WrapperCompile,
+we only tag the "base error constructor" PirCompile:
+
+```
+instance ErrorCode PirError where
+   errorCode PirCompile {} = 9997
+   errorCode (WrapperTC e) = errorCode e
+   errorCode (WrapperParse e) = errorCode e
+```
+
+A Plutus sub-project that wants to throw an error, must depend on this package `plutus-common`.
+To aid in defining an instance for a brand-new (uncategorized) error or added error-dataconstructors,
+the Plutus developer can make use (but not cabal-depend upon) of the
+mega-package plutus-errors to "guess-pick" an error-code that is not currently in use
+by the whole codebase, by running
+
+```
+> cabal run plutus-errors:next-error.
+An error code that is not currently used is: 9998
+```
+
+After defining/extending this errorcode instance, the Plutus developer must navigate to the megapackage and
+confirm these new errorcodes by adding all newly-introduced base-error constructors
+to the list of all-project errors `plutus-errors/src/Errors.hs`. The TH code of `plutus-errors`
+will make sure there are not duplicates, and running the `cabal haddock plutus-errors`
+will build the documentation of all (categorized) errors.
+
+When an error is deprecated (does not trigger anymore) and (some of) its dataconstructors has been removed,
+and in case the error is "exposed" to the public, then it is required that its "deprecated" constructors
+be "moved" and listed/errorCoded under the umbrella datatype `plutus-errors:Errors.DeprecatedErrors`.
+The reason for this is to document/keep track of deprecated errors and not *re*-pick "old" error-codes.
+
+Currently all errors among the project are placed into one big pile of error-codes. We later
+might use sub-groups of error-codes with specific word ranges, e.g. (PIR : 0000 - 0100, PLC: 0100 - 0200, etc), which then would require
+to put into use the "wrapper-constructors" of our error-grouppings.
```

Comment:
I still think we should do this, and not with "reserved ranges", but just with prefixes. e.g. `PLC001` is a PLC error, `PTX001` is a Plutus Tx error. That way you can do the non-overlapping checks locally per-package, and also build a directory of errors per-package.

Reply:
Actually after thinking more about it, I neither like sub-groups like PLC001 or PTX001, nor errocode ranges. From a user perspective I think the "one big bag" of errorcodes works better. There is not going to be a confusion between PLC001 or PTX001 or PLL0001 , etc... @effectfully  what is your opinion?

Reply:
I don't think PLC001 and PTX001 are confusing. They're different strings! As a user, I'm going to treat this mostly as an opaque string that I'm going to google, so ti doesn't matter that much.

Mainly I think it'll make it easier for us to maintain them.

Reply:
"PLC001" is easier to google than "0001" or "1", so I like this subcategories idea.